In [2]:
import pandas as pd
df=pd.read_csv('/Users/cyrilmahfouz/Desktop/SEF/Week2/Pandas_Rfinal/Telco-Customer-Churn.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


<h3><font color='red'>Introduction</font></h3>

In [3]:
df.head()


customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

To start we are looking at a dataframe representing the customers in a telecominication company .This dataframe contains information about 7043 customer.Furthermore we can detect that all of these columns revolves around one target value and that is whether the customer churn or not.<br>
Let's begin by identifying the interesting columns and how could they possibly relate to the churning of the customer. The logical links that affect the prices are:<br>
1. The tenure.<br>
2. The age range (Senior or not)
3. Whether or not the customer have dependents or not.(Partners).<br>
4. What is the type of service that he is using.<br>
5. The type of contract that he is enrolled with.<br>
6. The total charges that he has to pay. 

<h3><font color='red'>Anazlyze</font></h3>

To begin with let's understand the meaning of churn and why it is that important. The churn column represents whether the custumer bail on the company or not. So we could also get from this column if the customer is loyal towards the company.<br>
After understandig this, lets kick it by getting the total number of churn and analyzing the tenure of each customer.

In [4]:
total_churn=df.loc[df.Churn=='Yes']['customerID'].count()
print("the total number of churn is: ",total_churn)

the total number of churn is:  1869


<h4><b>The tenure</b></h4>

The tenure of a customer refers to the amount of time a customer has been using or subscribed to a particular service.Lets take a deeper look into it:

In [5]:
avrg_tenure=df['tenure'].mean().round()
print("the average tenure duration is: ",avrg_tenure)
print("the maximum tenure duration is: ",df['tenure'].max())
print()

#lets group them into 2 groups: 
df_=df.copy()
df_['above_avrg_tenure']=df_['tenure'].apply(lambda x:1 if x>avrg_tenure else 0)


print("the number of customer above avrg tenure: ",df_.loc[df_['above_avrg_tenure']== 1]['customerID'].count())
print("the number of customer lower than avrg tenure: ",df_.loc[df_['above_avrg_tenure']== 0]['customerID'].count())
print()

print("number of churn for customer having above avrg tenure: ",df_.loc[(df_['above_avrg_tenure'] == 1) & (df_['Churn'] == 'Yes')]['Churn'].count())
print('number of churn for customer having lower avrg tenure: ',df_.loc[(df_['above_avrg_tenure'] == 0) & (df_['Churn'] == 'Yes')]['Churn'].count())


the average tenure duration is:  32.0
the maximum tenure duration is:  72

the number of customer above avrg tenure:  3268
the number of customer lower than avrg tenure:  3775

number of churn for customer having above avrg tenure:  409
number of churn for customer having lower avrg tenure:  1460


First of all we started the analysis by getting the average tenure duration of the customers and we got the number 32.<br>Then we proceeded by grouping the customers into 2 groups. The first group represents the customer with an above average tenure duration and the second group represents customer with a lower average tenure duration.<br>
<br>
It appears that the number of customer that are in the first group are 3268 person and the number of customer that are in the second group are 3775 person;thise 2 numbers are very close to each other which is great for determining whether or not the group has a tendency to churn.<br>
<br>
We carry on by checking both groups number of churn customer and here is where it gets interesting.Even though the number of customer in each group is close, we remark that the number of customer that churn and have  an above avrg tenure is 409 and the number of churn for customer having lower average tenure is 1460<br>
<br><font color='"#FFCCCB"'>In conclusion, it's evident that customers with above-average tenure show a lower churn rate compared to customers with below-average tenure. This observation highlights the significant impact of tenure on customer churn within the provided dataset.</font><br>
<br>

<h4><b>Age range</b></h4>

In [6]:
senior=df_.loc[df_.SeniorCitizen==1]['customerID'].count()
junior=7043-senior
print("the number of senior citizen customer is: ",senior)
print("the number of junior citizen customer is: ",junior)
print()

senior_churn=df_.loc[(df_.SeniorCitizen==1) & (df_.Churn=='Yes')]['customerID'].count()
junior_churn=df_.loc[(df_.SeniorCitizen==0) & (df_.Churn=='Yes')]['customerID'].count()
print("the total number of senior that bailed: ",senior_churn)
print("the total number of junior thay bailed: ",junior_churn)
print()


perc_senior_custnb=(senior/7043)*100
perc_junior_custnb=100-perc_senior_custnb
print("the percentage of senior citizen that are customers is: ",perc_senior_custnb.round(2))
print("the percentage of junior citizen that are customers is: ",perc_junior_custnb.round(2))
print()

perc_senior_bail=(senior_churn/total_churn)*100
perc_junior_bail=100-perc_senior_bail
print("the percentage of the customer that bailed and is a senior citizen is: ",perc_senior_bail.round(2))
print("the percentage of the customer that bailed and is a junior citizen is: ",perc_junior_bail.round(2))

the number of senior citizen customer is:  1142
the number of junior citizen customer is:  5901

the total number of senior that bailed:  476
the total number of junior thay bailed:  1393

the percentage of senior citizen that are customers is:  16.21
the percentage of junior citizen that are customers is:  83.79

the percentage of the customer that bailed and is a senior citizen is:  25.47
the percentage of the customer that bailed and is a junior citizen is:  74.53


Concerning the age range of the customer, we began by grouping the data into 2 groups;senior citizen and junior citizen<br>
<br>The number of senior citizen present in our data is 1142 and the number of junior citizen is 5901.These numbers seems far from each other but there is something intreaguing about them.<br>
<br>Even though the percentage representing the senior customer is only 16.21% ,the percentage representing the customer that are senior and became unloyal is 25% which is a quarter of the churn column.<br>
<br><font color='"#FFCCCB"'>So, even though the number of senior citizen is low, the percentage of bail is not.It could be affected by a lot of circumstances and maybe somebody rely on them for support. Lets tackle the dependencie section<br>
<br>


<h4><b>Dependencie and partnership</b></h4>

In [7]:
total_sen_dep=df.loc[(df_['SeniorCitizen']==1) & (df_['Dependents']=='Yes')]
total_sen_not_dep=df.loc[(df_['SeniorCitizen']==1) & (df_['Dependents']=='No')]

print("total number of senior citizen that have people depending on him: ",total_sen_dep['customerID'].count())
print("the senios that doesnt have people depending on them: ",total_sen_not_dep['customerID'].count())
print()

print("senior that have people depending on them and bailed: ",total_sen_dep.loc[total_sen_dep['Churn']=='Yes']['customerID'].count())
print("senior that doesn't have people depending on them and bailed: ",total_sen_not_dep.loc[total_sen_not_dep['Churn']=='Yes']['customerID'].count())
print()

perc_sen_dep=(22/91)*100
perc_not_sen_dep=(454/1051)*100

print("the percentage of people that bailed and have dependents: ",round(perc_sen_dep,2))
print("the percentage of people that bailed and doesn't have dependents: ",round(perc_not_sen_dep,2))

total number of senior citizen that have people depending on him:  91
the senios that doesnt have people depending on them:  1051

senior that have people depending on them and bailed:  22
senior that doesn't have people depending on them and bailed:  454

the percentage of people that bailed and have dependents:  24.18
the percentage of people that bailed and doesn't have dependents:  43.2


To continue with the analyze that we did before we can conclude that the senior citizens are somewhat disloyal to the company.Because 43.2% senior citizen that didn't have dependents departed.<br>
Now let's proceed by checking the other part of the customers;the juniors

In [8]:
total_jun_dep=df.loc[(df_['SeniorCitizen']==0) & (df_['Dependents']=='Yes')]
total_jun_not_dep=df.loc[(df_['SeniorCitizen']==0) & (df_['Dependents']=='No')]

print("total number of junior citizens that have people depending on them: ",total_jun_dep['customerID'].count())
print("total number of junior citizens that doesnt have people depending on them: ",total_jun_not_dep['customerID'].count())
print()

print("junior that have people depending on them and bailed: ",total_jun_dep.loc[total_jun_dep['Churn']=='Yes']['customerID'].count())
print("junior that doesn't have people depending on them and bailed: ",total_jun_not_dep.loc[total_jun_not_dep['Churn']=='Yes']['customerID'].count())
print()

perc_jun_dep=(304/2019)*100
perc_not_jun_dep=(1089/3882)*100

print("the percentage of people that bailed and have dependents: ",round(perc_jun_dep,2))
print("the percentage of people that bailed and doesn't have dependents: ",round(perc_not_jun_dep,2))

total number of junior citizens that have people depending on them:  2019
total number of junior citizens that doesnt have people depending on them:  3882

junior that have people depending on them and bailed:  304
junior that doesn't have people depending on them and bailed:  1089

the percentage of people that bailed and have dependents:  15.06
the percentage of people that bailed and doesn't have dependents:  28.05


From this data,we can clearly see that the most loyal customers belong to the category of junior that have dependents;only 15.06% of them stopped there contract with the company.We can clearly see a diffrence in the percentages between the senior citizens that doesn't have dependents and the junior citizens that shares the same column. For the senior we had a percentage of 43.2% nearly half of them left the company, and for the junior only 28.05%.<br>
So both age and dependencie are factors that could determine whether or not this client will ditch the company.But this doesnt end here, let's check the partnership

In [17]:
print("partner+churn: ",df_.loc[(df_['Partner']=='Yes') & (df_['Churn']=='Yes')]['customerID'].count())
print("no partner+churn: ",df_.loc[(df_['Partner']=='No') & (df_['Churn']=='Yes')]['customerID'].count())
print()




partner+churn:  669
no partner+churn:  1200



Here we can see a strong relation that affect the churn column.We can clearly see that the customers with no partner are more likely to end up unloyal than the customers that have a partner.

<h4><b>Services</b></h4>

While reading the data, we can notice two types of service;phone services and internet services.Lets begin by splitting them into 2 groups.

In [33]:
#Phone services
only_phone=df_.loc[(df_['PhoneService']=='Yes') & (df_['InternetService']=='No')]
perc_phone=(1526/7043)*100
print("only phone service customers: ",only_phone.customerID.count())
print("percentage of phone service customer only: ",round(perc_phone,2))
print()

one_line_cust=only_phone.loc[only_phone['MultipleLines']=='No']
print("the number of customers that have 1 line: ",one_line_cust.customerID.count())
print("nb of customers that have multiple lines: ",only_phone.loc[only_phone['MultipleLines']=='Yes'].customerID.count())
print()

one_line_churn=one_line_cust.loc[one_line_cust.Churn=="Yes"]
mult_line_churn=only_phone.loc[(only_phone['MultipleLines']=='Yes') & (only_phone['Churn']=="Yes")]
print("1 line + churn: ",one_line_churn.customerID.count())
print("multiple lines + churn: ",mult_line_churn.customerID.count())
print()

perc_one_churn=(103/1184)*100
perc_mult_churn=(10/342)*100
print("percentage of one line +churn: ",round(perc_one_churn,2))
print("percentage of multiple line + churn: ",round(perc_mult_churn,2))



only phone service customers:  1526
percentage of phone service customer only:  21.67

the number of customers that have 1 line:  1184
nb of customers that have multiple lines:  342

1 line + churn:  103
multiple lines + churn:  10

percentage of one line +churn:  8.7
percentage of multiple line + churn:  2.92


We can clearly identify that the phone service is satisfying to the customers because the percentage of churn of the single line is 8.7% and the percentage of churn for the multiple line service is only 2.92%.<br>
But we are still talking about the customers that only uses the phone service.Lets continue by analyzing the internet only customer:

In [47]:
#internet cust only
int_cust=df_.loc[(df_['PhoneService']=='No') & ((df_['InternetService']=='DSL') | (df_['InternetService']=="Fiber optic"))]

perc_int=(int_cust.customerID.count()/7043)*100
print("only internet service customers: ",int_cust.customerID.count())
print("percentage of internet service customer only: ",round(perc_int,2))
print()

dsl_cust=df_.loc[(df_['PhoneService']=='No') & (df_['InternetService']=='DSL')]
fiber_cust=df_.loc[(df_['PhoneService']=='No') & (df_['InternetService']=='Fiber optic')]

print("total nb of churn for internet cust: ",int_cust.loc[int_cust['Churn']=="Yes"].customerID.count())
print("nb of churn for DSL customers: ",dsl_cust.loc[dsl_cust['Churn']=="Yes"].customerID.count())
print("nb of churn for fiber customer: ",fiber_cust.loc[fiber_cust['Churn']=='Yes'].customerID.count())



only internet service customers:  682
percentage of internet service customer only:  9.68

total nb of churn for internet cust:  170
nb of churn for DSL customers:  170
nb of churn for fiber customer:  0


Before we go deeper into the internet services lets have a recap.<br>
So first we splitted the customers into 2 parts: the first part are the customers that only have phone services, and the second part are the customers that only have internet access.<br>
After studying both group we got that the phone customers represents 21.67% of all the customers and the internet only group represented 9.68% of them.<br>
Even though the number of phone customer is bigger we found out that only 113 of them didn't continue with the company which is smaller than the one for the internet customer that is 170;ut these 170 churn only comes from DSL users the fiber customer are loyal.<br>
<br>
We can remark that the phone service is much more loyal to the company maybe because the service is well provided and people are happy with it.